openedx/core/djangoapps/schedules/management/commands/__init__.py

```diff
@@ -45,6 +51,10 @@ def handle(self, *args, **options):
             num_days = (7 * num_weeks) + 1
             self.offsets = range(-7, -num_days, -7)
 
+        check_completion = options.get('check-completion')
+        if check_completion:
+            self.check_completion = check_completion
+
```

Comment:
So, this variable gets set on the `Command` object, but since it doesn't get passed through to the `enqueue` function, it doesn't get passed on to the task itself.

Here's the diff to do this, which also combines all changes noted in my other comments.

```diff
diff --git a/openedx/core/djangoapps/schedules/management/commands/__init__.py b/openedx/core/djangoapps/schedules/management/commands/__init__.py
index b60219bd9a..12f1f029fd 100644
--- a/openedx/core/djangoapps/schedules/management/commands/__init__.py
+++ b/openedx/core/djangoapps/schedules/management/commands/__init__.py
@@ -19,7 +19,6 @@ class SendEmailBaseCommand(PrefixedDebugLoggerMixin, BaseCommand):
     # An iterable of day offsets (e.g. -7, -14, -21, -28, ...) that defines the days for
     # which emails are sent out, relative to the 'date' parameter
     offsets = range(-7, -77, -7)
-    check_completion = False
 
     def add_arguments(self, parser):
         parser.add_argument(
@@ -39,7 +38,8 @@ class SendEmailBaseCommand(PrefixedDebugLoggerMixin, BaseCommand):
         )
         parser.add_argument(
             '--check-completion',
-            deafult=False,
+            default=False,
+            action='store_true',
             help='Check if the student has completed atleast one block'
         )
 
@@ -51,10 +51,6 @@ class SendEmailBaseCommand(PrefixedDebugLoggerMixin, BaseCommand):
             num_days = (7 * num_weeks) + 1
             self.offsets = range(-7, -num_days, -7)
 
-        check_completion = options.get('check-completion')
-        if check_completion:
-            self.check_completion = check_completion
-
         current_date = datetime.datetime(
             *[int(x) for x in options['date'].split('-')],
             tzinfo=pytz.UTC
@@ -65,14 +61,16 @@ class SendEmailBaseCommand(PrefixedDebugLoggerMixin, BaseCommand):
         self.log_debug(u'Running for site %s', site.domain)
 
         override_recipient_email = options.get('override_recipient_email')
-        self.send_emails(site, current_date, override_recipient_email)
+        check_completion = options.get('check_completion')
+        self.send_emails(site, current_date, override_recipient_email, check_completion)
 
-    def enqueue(self, day_offset, site, current_date, override_recipient_email=None):
+    def enqueue(self, day_offset, site, current_date, override_recipient_email=None, check_completion=False):
         self.async_send_task.enqueue(
             site,
             current_date,
             day_offset,
             override_recipient_email,
+            check_completion,
         )
 
     def send_emails(self, *args, **kwargs):
diff --git a/openedx/core/djangoapps/schedules/management/commands/send_course_update.py b/openedx/core/djangoapps/schedules/management/commands/send_course_update.py
index bcbbda82e2..43225caefd 100644
--- a/openedx/core/djangoapps/schedules/management/commands/send_course_update.py
+++ b/openedx/core/djangoapps/schedules/management/commands/send_course_update.py
@@ -17,4 +17,3 @@ class Command(SendEmailBaseCommand):
     help = dedent(__doc__).strip()
     async_send_task = ScheduleCourseUpdate
     log_prefix = 'Course Update'
-    check_completion = False
diff --git a/openedx/core/djangoapps/schedules/resolvers.py b/openedx/core/djangoapps/schedules/resolvers.py
index 2030164b52..a9958e661b 100644
--- a/openedx/core/djangoapps/schedules/resolvers.py
+++ b/openedx/core/djangoapps/schedules/resolvers.py
@@ -58,6 +58,8 @@ class BinnedSchedulesBaseResolver(PrefixedDebugLoggerMixin, RecipientResolver):
                         org_list or strictly include (False) them (default: False)
         override_recipient_email -- string email address that should receive all emails instead of the normal
                                     recipient. (default: None)
+        check_completion -- boolean indicating whether the user needs to have completed at least 1 course block before
+                            any scheduled emails will be sent. (default: False)
 
     Static attributes:
         schedule_date_field -- the name of the model field that represents the date that offsets should be computed
@@ -382,19 +384,17 @@ class CourseUpdateResolver(BinnedSchedulesBaseResolver):
         )
 
         template_context = get_base_template_context(self.site)
+        check_completion = (self.check_completion and
+            completion_waffle.waffle().is_enabled(completion_waffle.ENABLE_COMPLETION_TRACKING))
         for schedule in schedules:
             enrollment = schedule.enrollment
             course = schedule.enrollment.course
             user = enrollment.user
-            course_id = schedule.enrollment.course_id
-            course_key = CourseKey.from_string(course_id)
-            if self.check_completion:
-                if completion_waffle.waffle().is_enabled(completion_waffle.ENABLE_COMPLETION_TRACKING):
-                    last_block_completed = BlockCompletion.get_latest_block_completed(user, course_key)
-                    if not last_block_completed:
-                        week_num = 1
-                        self.day_offset = 1
-                        continue
+            course_key = schedule.enrollment.course_id
+            if check_completion:
+                last_block_completed = BlockCompletion.get_latest_block_completed(user, course_key)
+                if not last_block_completed:
+                    continue
 
             try:
                 week_highlights = get_week_highlights(user, enrollment.course_id, week_num)
diff --git a/openedx/core/djangoapps/schedules/tasks.py b/openedx/core/djangoapps/schedules/tasks.py
index e3ff5631f0..628279771c 100644
--- a/openedx/core/djangoapps/schedules/tasks.py
+++ b/openedx/core/djangoapps/schedules/tasks.py
-            course_key = CourseKey.from_string(course_id)
-            if self.check_completion:
-                if completion_waffle.waffle().is_enabled(completion_waffle.ENABLE_COMPLETION_TRACKING):
-                    last_block_completed = BlockCompletion.get_latest_block_completed(user, course_key)
-                    if not last_block_completed:
-                        week_num = 1
-                        self.day_offset = 1
-                        continue
+            course_key = schedule.enrollment.course_id
+            if check_completion:
+                last_block_completed = BlockCompletion.get_latest_block_completed(user, course_key)
+                if not last_block_completed:
+                    continue
 
             try:
                 week_highlights = get_week_highlights(user, enrollment.course_id, week_num)
diff --git a/openedx/core/djangoapps/schedules/tasks.py b/openedx/core/djangoapps/schedules/tasks.py
index e3ff5631f0..628279771c 100644
--- a/openedx/core/djangoapps/schedules/tasks.py
+++ b/openedx/core/djangoapps/schedules/tasks.py
@@ -85,7 +85,7 @@ class ScheduleMessageBaseTask(LoggedTask):
         LOG.info(cls.log_prefix + ': ' + message, *args, **kwargs)
 
     @classmethod
-    def enqueue(cls, site, current_date, day_offset, override_recipient_email=None):
+    def enqueue(cls, site, current_date, day_offset, override_recipient_email=None, check_completion=False):
         current_date = resolvers._get_datetime_beginning_of_day(current_date)
 
         if not cls.is_enqueue_enabled(site):
@@ -101,6 +101,7 @@ class ScheduleMessageBaseTask(LoggedTask):
                 day_offset,
                 bin,
                 override_recipient_email,
+                check_completion,
             )
             cls.log_info(u'Launching task with args = %r', task_args)
             cls().apply_async(
@@ -115,7 +116,7 @@ class ScheduleMessageBaseTask(LoggedTask):
         return False
 
     def run(
-        self, site_id, target_day_str, day_offset, bin_num, override_recipient_email=None,
+        self, site_id, target_day_str, day_offset, bin_num, override_recipient_email=None, check_completion=False,
     ):
         site = Site.objects.select_related('configuration').get(id=site_id)
         with emulate_http_request(site=site):
@@ -128,6 +129,7 @@ class ScheduleMessageBaseTask(LoggedTask):
                 day_offset,
                 bin_num,
                 override_recipient_email=override_recipient_email,
+                check_completion=check_completion,
             ).send(msg_type)
 
     def make_message_type(self, day_offset):

```

Reply:
👍